lotus-soup/testkit/role_miner_sync.go

```diff
@@ -0,0 +1,169 @@
+package testkit
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/testground/sdk-go/sync"
+)
+
+// MinerSyncTopic is the topic where miner events are exchanged.
+var MinerSyncTopic = sync.NewTopic("miner_sync", &MiningSyncMsg{})
+
+// MiningSyncMsg represents a miner event.
+type MiningSyncMsg struct {
+	Addr     address.Address
+	Produced bool // whether this miner produced a block.
+}
+
+// SynchronizeMiner synchronizes this miner with all other miners in this test,
+// so that they're mining in perfect synchrony according to a globally-
+// synchronized epoch clock.
+//
+// This function _receives_ "epoch start" signals and _emits_ "epoch end"
+// signals:
+//
+//  * an epoch start signal attempts to locally mine a block, and publishes
+//    a message on a sync topic indicating that we're done mining, and whether
+//    we produced a block or not.
+//  * all miners wait until their HEAD advances to the tipset that includes
+//    all blocks that were produced in this epoch.
+//  * once that happens, we emit an epoch end signal, which the caller can
+//    interpret as a signal to advance the clock and begin a new epoch.
+func (m *LotusMiner) SynchronizeMiner(ctx context.Context, epochStart <-chan abi.ChainEpoch, epochEnd chan<- abi.ChainEpoch) {
+	var (
+		minerCnt = m.t.IntParam("miners")
+		produced = make(map[address.Address]struct{}, minerCnt) // addresses that produced a block in this epoch.
+
+		epoch = abi.ChainEpoch(0)
+		done  []address.Address
+
+		signalCh = make(chan *MiningSyncMsg, 128)
+		sub      = m.t.SyncClient.MustSubscribe(ctx, MinerSyncTopic, signalCh)
+	)
+
+	// obtain our address so we can broadcast it in signal messages.
+	actorAddr, err := m.MinerApi.ActorAddress(ctx)
+	if err != nil {
+		panic("failed to obtain actor address: " + err.Error())
+	}
+
+	// function to mine a single round.
+	mineRound := func(epoch abi.ChainEpoch) (produced bool) {
+		// mining can fail, so we perform 100 retries.
```

Comment:
I don't really know why it can fail. I expanded a little bit, but was unable to explain under which circumstances this could happen. @yusefnapora, do you recall what the error said when this happened?